.NET SDK

ckpearson opened the pull request:
Addresses #28 

Provides a core .NET 9 Class Library that:

- Has all the core types and events
- Has an agent abstraction and implementations for `IChatClient` from `Microsoft.Extensions.AI` that:
  - Support frontend & backend tool calling
  - Support shared agentic state flows
- Has basic helper integration for hosting agents via ASP.NET endpoints

Also an initial stab at providing updates to the documentation site with enough information to get started using it.

Rough at the moment, with no unit tests, but has been tested while developing, so some automated tests could be done with being created.

Includes an example ASP.NET project using the class library to host a few agents.

ckpearson committed:
Core message and event types implemented, earliest work of the AbstractAgent implementation and basic echo agent for testing.

ckpearson committed:
Significant tidy up, introduced new agent constructs and fleshed out the ChatClientAgent to proxy over an LLM via IChatClient.

Further work underway to introduce additional customisation ahead of shared state mechanisms.

ckpearson committed:
fix: update user field names for clarity in output

ckpearson committed:
Further tweaks to support not tying system message overwriting solely into the context inclusion mechanism.

ckpearson committed:
Initial work on scaffolding docs for the .NET SDK

ckpearson committed:
refactor: improve formatting and structure in EchoAgent and add MapAgentEndpoint extension method

ckpearson committed:
Further work on .NET SDK documentation:

- Agents
- Overview
- ASP.NET integration

ckpearson committed:
fix: change default value of IncludeContextInSystemMessage to false

ckpearson committed:
Enhance .NET SDK Documentation

- Updated the .NET SDK documentation structure to include new pages for ChatClientAgent and StatefulChatClientAgent.
- Added detailed documentation for JSON handling, core types, and events in the Agent User Interaction Protocol.
- Improved the organization of the .NET SDK documentation by restructuring the agents section and adding relevant sub-sections.
- Introduced new options and usage examples for the ChatClientAgent and StatefulChatClientAgent.
- Enhanced the ASP.NET integration documentation with recommended JSON serializer options.

ckpearson committed:
fix: add ConfigureAwait(false) to asynchronous calls that aren't the ASP.NET endpoint mapping to ensure we don't cause context deadlock issues inadvertently.

Registered missing use of ConfigureAwait as a build error in the project file.

ckpearson committed:
feat: add options for backend tool call emission and state function handling in ChatClientAgent and StatefulChatClientAgent

ckpearson commented:
Made further tweaks which provides greater control, and now supports notifying the frontend about backend tool calls to power generative UI - albeit by sending message snapshots which I'm not 100% convinced is the correct approach.

ckpearson commented on dotnet-sdk/AGUIDotnet/Agent/StatefulChatClientAgent.cs:
```diff
@@ -0,0 +1,200 @@
+                }
+            }
+        }
+        catch (JsonException)
+        {
+
```
This is deliberate, for now it's just to avoid failure to extract the context causing an actual problem, it just swallows the exception, but we could perhaps surface it somehow so the consumer decides what behaviour to exhibit.

tylerslaton commented:
As discussed during our synchronous review, there are few things needed prior to merging this but it is really close.

- [ ] The README should match the README's of the other SDK's, pretty small change
- [ ] We will need tests to ensure we don't have regressions introduced in PR's

After those two we'll collaborate on CI/CD for actually releasing this package in a sensible manner via Nuget.

ckpearson committed:
Add AGUIDotnet.Tests project and implement initial test cases

- Created AGUIDotnet.Tests project with necessary dependencies.
- Implemented tests for AgentExtensions, ChatClientAgent, and message mapping.
- Added tests for frontend tool invocation and tool message mapping.
- Updated AGUIDotnet solution to include the new test project.
- Enhanced AgentExtensions to ensure proper channel completion.
- Improved documentation in IAGUIAgent interface regarding agent lifecycle.
- Removed unused HTTP request file from AGUIDotnetWebApiExample.
- Updated README.md to include links to official documentation.

ckpearson committed:
feat: implement TestStreamingChatClient for simulating streaming responses

ckpearson commented:
Sorry for the delay, there should be more broad test coverage of the basics, and the README has been updated.

GreenShadeZhang commented:
May I ask if this PR is being processed?

NathanTarbert commented:
Hi @GreenShadeZhang, we are waiting for this SDK to be finalized before we can fully review it.

brmdbr commented:
Hopefully this can be merged soon!

ckpearson commented:
> Hopefully this can be merged soon!

Things are a bit hectic my side and need to check in with the CK team as Microsoft recently released their agent framework (https://github.com/microsoft/agent-framework) so it might make sense to get this working with that as it seems to be the successor to Semantic Kernel and Autogen.

danroth27 commented:
> Things are a bit hectic my side and need to check in with the CK team as Microsoft recently released their agent framework (https://github.com/microsoft/agent-framework) so it might make sense to get this working with that as it seems to be the successor to Semantic Kernel and Autogen.

Hi @ckpearson. We're now exploring adding AG-UI support as part of the Microsoft Agent Framework: https://github.com/microsoft/agent-framework/issues/896. It would be great if we could sync up and coordinate on our efforts.
